rfcs/20201121-keras-model-fit-ps.md

```diff
@@ -0,0 +1,508 @@
+
+
+```
+cluster_resolver = tf.distribute.cluster_resolver.TFConfigClusterResolver()
+server = tf.distribute.Server(
+    cluster_resolver.cluster_spec(),
+    job_name=cluster_resolver.task_type,
+    task_index=cluster_resolver.task_id,
+    protocol=cluster_resolver.rpc_layer or "grpc",
+    start=True)
+server.join()
+```
+
+
+
+### Evaluation
+
+If `validation_data` is provided, and certain conditions are satisfied, `model.fit` also runs evaluation via `model.evaluate` API every epoch, in an train-evaluate alternating manner. As described above, at this time, only coordinator is used for `model.evaluate` evaluation, and we plan to extend this to worker-distributed evaluation when visitation guarantee is supported.
+
+In addition to the built-in evaluation `model.fit` provides, sidecar evaluation is also supported with the `SidecarEvaluator` [API](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/distribute/sidecar_evaluator.py). User is expected to kick start an additional task `evaluator`, in which the python program runs a `SidecarEvaluator` as follows:
+
+
+```
+model = tf.keras.models.Sequential(...)
+model.compile(metrics=tf.keras.metrics.SparseCategoricalAccuracy(...)
+data = tf.data.Dataset.from_tensor_slices(...)
+
+SidecarEvaluator(
+    model=model,
+    data=data,
+    checkpoint_dir='/tmp/checkpoint_dir',  # dir for training-saved checkpoint
+    steps=None,  # Eval until dataset is exhausted
+    log_dir='/tmp/log_dir',
+    max_evaluations=None  # The evaluation needs to be stopped manually
+).start()
+```
+
+
+`SidecarEvaluator` periodically loads the checkpoint files saved by the training counterpart, as long as the checkpoint captures the `model` and `optimizer` objects. As part of full integration with `model.fit` workflow, we propose to extend `SidecarEvaluator` to 
+
+
+
+*   also accept the checkpoint files saved by `ModelCheckpoint` callback for periodic evaluation.
+*   accept arbitrary callbacks to be used in its internal `model.evaluate` call
+
+Users will then have two evaluation schemes to choose from: alternating evaluation, or sidecar evaluation, or both, depending on their needs.
+
+
+### Fault tolerance
+
+There are two goals of fault tolerance in multi-worker training:
+
+
+
+*   Task preemption or unavailability is handled by the program or cluster, without being reported as failure
+*   The training progress can withstand task preemption and subsequent recovery 
+
+The first goal is fulfilled by the failure handling mechanism provided by the `ClusterCoordinator` library, where worker unavailability is handled within the program, and PS/coordinator unavailability is handled by the cluster’s auto-restart mechanism (by exiting with an exit code that the cluster management recognizes). From the user's perspective, they can leverage proposed `handle_restarable_error` context manager:
+
+
+```
+with handle_restartable_error():
```

Comment:
Is this API part of the proposal?

Reply:
Updated this part. We expect users to use try-except.